### PR TITLE
App strings reworked

### PR DESCRIPTION
**Pull request by comradekingu**

### CODE REVIEW

File: app/aboutdialog.cpp
Comment (comradekingu) [inline app/aboutdialog.cpp:74]:
```suggestion
            tr("%1 is MIT licensed."), // %1

```

File: app/aboutdialog.cpp
Comment (comradekingu) [inline app/aboutdialog.cpp:82]:
```suggestion
        QStringLiteral("<p>%1</p>").arg(tr("The MIT license grants you these freedoms.")),

```

File: app/aboutdialog.cpp
Comment (comradekingu) [inline app/aboutdialog.cpp:81]: This is very misleading. Only the license text needs to be retained. The MIT license is not copyleft.

File: app/aboutdialog.cpp
Comment (comradekingu) [inline app/aboutdialog.cpp:111]: Not sure if this should be title case. Is it a headline?

File: app/graphicsview.cpp
Comment (comradekingu) [inline app/graphicsview.cpp:276]: https://en.wikipedia.org/wiki/Media_type or the data within it?

File: app/aboutdialog.cpp
Comment (BLumia) [reply app/aboutdialog.cpp:81]: Yeah The MIT license is not copyleft, but it did guarantee users this freedom as long as the user gets the software and source code, and for the copy of software/source code, the freedom (guarantee by MIT license) is not permitted to take away by anybody else, so IMO this is okay to leave it here. Is that correct?

I'm not sure if I understand this correctly, let me know if it's wrong.

File: app/graphicsview.cpp
Comment (BLumia) [inline app/graphicsview.cpp:261]: Not quite sure about the actual differences between URI and URL, since we (including Qt itself) are using QUrl all the time, maybe keep using "URL" could be better?

File: app/graphicsview.cpp
Comment (BLumia) [reply app/graphicsview.cpp:276]: It means the MIME-type is not supported.

**Review by BLumia — COMMENTED** [review]
I'm not a native English speaker, sorry for the issues from the original English strings. I'll get it merged after these comments get resolved.

Thanks for the help!

File: app/aboutdialog.cpp
Comment (BLumia) [reply app/aboutdialog.cpp:111]: image
Yes, the formats are the same between all other tabs.

File: app/mainwindow.cpp
Comment (BLumia) [inline app/mainwindow.cpp:478]: This shouldn't get changed since the text is for `LANG=C` which will only use ASCII encoding, not for something similar to `en_US.UTF-8`. See the [`locale` manpage](https://man.openbsd.org/locale.1), [`setlocale` manpage](https://man7.org/linux/man-pages/man3/setlocale.3.html), [debian i18n page](https://www.debian.org/doc/manuals/debian-reference/ch08.en.html#_basics_of_encoding) for details.

If you would like to use `…` for UI string displaying, consider adding something like `en` or `en_US` as a translatable language and do it there?

**Review by BLumia — CHANGES_REQUESTED** [review]
Oops sorry, I missed one issue which needs to be changed. See the comment for detail.

File: app/aboutdialog.cpp
Comment (BLumia) [inline app/aboutdialog.cpp:47]: `©` is also not a part of the ASCII character set, so in the original text for `LANG=C`, we prefer to use `(c)`.

**Review by BLumia — CHANGES_REQUESTED** [review]

File: app/aboutdialog.cpp
Comment (comradekingu) [inline app/aboutdialog.cpp:47]:
```suggestion
        tr("Copyright (c) 2020 %1").arg(QStringLiteral("<a href='https://github.com/BLumia'>@BLumia</a>")),

```

File: app/aboutdialog.cpp
Comment (comradekingu) [inline app/aboutdialog.cpp:111]:
```suggestion
        QStringLiteral("<h1 align='center'><b>%1</b></h1>").arg(tr("Third-party Libraries Used by %1")),

```

File: app/graphicsview.cpp
Comment (comradekingu) [reply app/graphicsview.cpp:261]: Ideally none of those, but will make it URL.

Comment (comradekingu) [inline app/graphicsview.cpp:261]:
```suggestion
            showText(tr("The file URL list is empty"));

```

File: app/graphicsview.cpp
Comment (comradekingu) [reply app/graphicsview.cpp:276]: OK, then this is good. MIME-type changed name to "media type" some time ago.

File: app/mainwindow.cpp
Comment (comradekingu) [reply app/mainwindow.cpp:478]: Would work if that is the translated language too, otherwise it fails the check between tripledot and ellipsis from source to translation.

Comment (comradekingu) [inline app/mainwindow.cpp:478]:
```suggestion
    QAction * toggleSettings = new QAction(tr("Configure..."));

```

File: app/aboutdialog.cpp
Comment (BLumia) [reply app/aboutdialog.cpp:81]: > This is very misleading. ... The MIT license is not copyleft.

[The four essential freedoms](https://www.gnu.org/philosophy/free-sw.html.en) doesn't mean the software should have a copyleft license. The MIT license actually grants users these rights (and MIT is also an [FSF-approved license](https://www.gnu.org/licenses/license-list.en.html#Expat)), so I think this text is not misleading at all and I prefer to retain this text if I understand correctly.

**Review by BLumia — CHANGES_REQUESTED** [review]
Other seems fine, one last comment need to be discussed. I'll leave my thought in that comment thread (and please let me know if I'm wrong). Will merge the PR once the last one is resolved :)

File: app/aboutdialog.cpp
Comment (comradekingu) [reply app/aboutdialog.cpp:81]: True, it is this part "Nobody is ever permitted to take it away." which is indistinguishable from copyleft. In some sense of the wording "it" means the freedoms, which can be "taken away" since it is permissible. The MIT license itself, if that is "it", can also effectively be taken away. I already think "guarantees" is strong, but I don't see how "Nobody is ever", which couldn't be stronger (and probably doesn't account for it entering the public domain?) etc. adds anything.

Comment (BLumia) [reply app/aboutdialog.cpp:81]: > In some sense of the wording "it" means the freedoms, which can be "taken away" since it is permissible.
> The MIT license itself, if that is "it", can also effectively be taken away.

That's not true, the MIT license allowed gives user those essential freedoms and there is no way to revoke the license which the user is already gotten (unless user abuse the license itself). If someone else takes a copy of this software and turn their forked version into non-free software, that won't affect the other users who get their Expat/MIT licensed copy, since the Expat/MIT license grants users the four essential freedoms, nobody is ever permitted to take it away if the user already gots their license which grants users these freedoms. (FSF won't approve the license if someone else could "take away" these freedoms)

I'm not sure why you think the freedoms can be taken away from users if the software is using a not copyleft license, If you think someday the copyright holder (I) can change the license and turn it into a non-free software, let's assume I did so, if the users can still find and read that text, it means that user already got an Expat/MIT licensed copy of this software, which means they still have the freedom to use the copy they got since the license grants the user these freedoms. So as you can see, even if I can do that will still not able take away the freedom from the users who already get a copy of the software (along with a valid Expat/MIT license), I think writing "Nobody is ever permitted to take it away" inside the software is okay.

Is that correct?

Comment (comradekingu) [reply app/aboutdialog.cpp:81]: "which the user is already gotten" isn't mentioned, and crucially is thing you have to already know about to make that work.
"Nobody is ever permitted to take it away" as it relates to "it", also applies for what _the source code_ is as it exists, not merely how you got it. The word "guarantee" alone is probably unwarranted, but certainly if it also covers this (without the disclaimer of distribution).

Similarly and furthermore,
"You can distribute changed versions of %1" doesn't hold true for
"Nobody is ever permitted to take it away."

I think the Freedoms are the ability to do these things. When those things are no longer possible, I don't think it matters that distribution is what granted the ability in the first place. Distribution as it relates to licensing is just a function of copyright. I don't care for copyright law, and unfortunately licenses have to be longer because the world of "abusing the license" isn't ideal. Having to deal in non-ideals, that is my take, and it is hard enough to convey.

"users who already get a copy of the software" I think _the software_ applies to what the software is, not what someone decided on as they distributed it to me. Yes, there needs to be a meaningful contribution to distribute it under a different license (I think), but that is semantics in terms of what user freedoms to the version of the software that matters.

As an example of the ambiguity, lets say you own the rights to some land. If I lease it for 10000 years, you don't own it less, but the difference between selling it to me and being able to do anything with the ownership vanishes in a way that doesn't warrant. "You keep owning your land just like before." It is true in some sense, but it isn't a good application of language unless you want to mislead.

Comment (BLumia) [reply app/aboutdialog.cpp:81]: > What the MIT license doesn't do is even grant, much less guarantee the additional copyleft.

But what the text in the application referred to is **the four essential freedoms** which provided by the Expat/MIT license. It's not related to copyleft at all.

> isn't mentioned, and crucially is the thing you have to already know about to make that work.

I don't think we should mention it again since if the user can read that text, obviously, the user already got their Expat/MIT licensed copy. The text itself is inside the application, not in somewhere like a homepage/official website. The text is intended to tell the user this software is free software that comes with a free (as in speetch) license (Expat/MIT) which grantee user freedom they already gotten.

> there is some ambiguity over what "it" means, but in either case it is shifty. "Freedom" as a whole can cover many concepts, so it could very well be all the above freedoms.

We are focus on the four essential freedoms, which is perfectly clear MIT license grantee users these freedoms.

> "You can distribute changed versions of %1" doesn't hold true for
> "Nobody is ever permitted to take it away."

Nobody is ever permitted to take it (for the user, the right to redistribute their changed version) away, that's always true. Elsewhere the original MIT licensed software won't be free software.

> I think the Freedoms are the ability to do these things. When those things are no longer possible, I don't think it matters that distribution is what granted the ability in the first place.

Yeah, and that's what I've said, users always have their right to do that if they got their Expat/MIT licensed copy, they even have their right to turn it into non-free software if needed, I'm not able to take their freedom away. But to make it perfectly clear, these rights don't mean the license should be copylefted.

> *I think* the software applies to what the software is

Since the text itself is inside the application, it refers to the copy they got. What you've said will make a scene only if I put that text in a general public place to let other people who haven't got their license copy. And also, since it's free software, even if I made my decision to turn it closed-source or something else, other people who got their free (as in speech) copy still have their rights to maintain a free libre fork.

The text is explaining how the license working and what did the license grants to the user. It's all about license text in a tab dedicated to licensing, of course, it's all related to licensing itself than covering the whole project management.

> "You keep owning your land just like before."

**That's how (free/libre) licenses work**. and that's why the explanation text meant to be related to the copy they got instead of refers to the whole project.

> It is true in some sense, but it isn't a good application of language unless you want to mislead.

A copylefted license adding extra limitation about how users can use the source code, but it's still true the license itself can only apply to the copy they got. It's not misleading at all. We are restricted to explain the license terms here, not about how the project will be in the future.

> There is a near zero chance it won't get even worse in translation.

The actual valid text in legal is still the original license text(which provided below the explanation text. If there is an error in a translation, such an error needs to be corrected. If you think the explain text may be wrong during translation, a disclaimer text can also be added, but since the original text is valid, I prefer to keep it here instead of removing it.

I don't think that could be an issue since Krita also using that text to tell users they have these freedoms. I take the idea to my project here since it generally applies to all free software, no matter it's using a copylefted license or a lax permissive license.

---------------

By the way, I can also explain why I choose a lax permissive license for this project instead of a copylefted license. The project itself is pretty simple and stupid, it provides users a way to open images. The project is intended to be tiny and small and won't have a bunch of features like other image management software, as [GNU also doesn't recommend to use a copylefted license for the small project](https://www.gnu.org/licenses/gpl-faq.en.html#WhatIfWorkIsShort), I choose Expat/MIT license to grant user more freedom than make the project copylefted.

Another thing is, if I made it GPL/LGPL, once other contributors contribute copyrightable source code into the codebase, I'll no longer have the right to redistribute the software to somewhere like Apple Store or other similar platforms due to the license limitation unless a CLA is required to be signed by contributors. That's really not what I want, if I made it Expat/MIT licensed, not only I could do this, other people can also do this with no strings attached as long as they are not abusing the license term.

In a conclusion, I prefer to keep the text here to tell users their rights grants by the license which comes with their copy of the free software, instead of removing it.

Comment (comradekingu) [reply app/aboutdialog.cpp:81]: The rights are still spelled out. They are not guaranteed, nor are they irrevocable. The changed version tells the user the freedoms are in the license, nothing more.

What the license is and what the FSF thinks about it is beside the point.

Comment (BLumia) [reply app/aboutdialog.cpp:81]: > They are not guaranteed, nor are they irrevocable

That's so untrue since it IS the license that guaranteed these freedoms and MIT is irrevocable (at least in the US, UK, EU, and mainland China AFAIK). Else it won't be able to be a free software license if the copyright holder can revoke someone else's license.

Please notice, relicense is a different thing, which doesn't affect the users who already got a valid license, software can be relicensed by the copyright holder doesn't mean the license which already grants to the user is not irrevocable.

**Comment by BLumia** [comment]
I tend to close this PR since it seems you do not fully understand the term. Free software license guarantees user their freedoms won't be taken by other people, and I'm intended to put the copywriting to tell users their rights. A lax permissive license is chosen because this is not a very large project which needs to be copylefted.

Also, some of the texts were changed/removed during updates, some of them were not changed but I'm planning to get a rewrite. I guess these issue will be resolved in coming updates :)

Anyway, still thanks for your patches :)

**Comment by comradekingu** [comment]
@BLumia 
The term isn't in dispute. What I am saying is the description would make sense if it was a copyleft license, where there actually is a guarantee. Pick the license you want obviously.

My issue is you are describing permissive licensing as if it were copyleft, that is misleading.
Whether the FSF approves it or what they recommend for small or big projects doesn't change matters.

It is equally true to say 

he MIT license guarantees everyone the freedom to stop development at any time to make versions that grant you no freedoms at all.

Therefore, "guarantees you", and "freedom" both being plural concept, is misleading.

Happy to rebase my work if you decide to ask someone that isn't me what they think about the claims.

**Comment by BLumia** [comment]
> What I am saying is the description would make sense if it was a copyleft license, ...

Since the copywriting clearly defined what the word freedom means here (the four essential freedoms), the description actually makes sense even if it is not a copyleft license, and that's why I refuse to remove it.

> ... a copyleft license, where there actually is a guarantee.

The license itself is a guarantee (of the freedoms mentioned in the copywriting I currently used), the license doesn't need to be copyleft to be legally valid.

Also, you may think a copyleft license is irrevocable according to the above discussion, which is also not true. Firstly, the copyright holder(s) can always change their license at any time if they like, even if they already chose a copylefted license, the *nmap* project is an example. Secondly, the (free software) license actually is irrevocable if one already got a valid license.

Copyleft is a form to guarantee *derivative work* will also be copyleft, they cannot guarantee the project itself in any form. A license is always and the only valid guarantee of the freedoms you got.

> (t)he MIT license guarantees everyone the freedom to stop development at any time to make versions that grant you no freedoms at all.

That's also freedom. Do you see? Everybody may have their own definition of *freedom* and it can be different, FSF, DFSG, and OSI both have different view about what actually is software freedom, and that's why the copywriting I provided also included the description of the four essential freedoms, which means I also give it a definition to avoid confusion.

I also do not want to introduce such a limitation that requires other people to close the source. I do think it's their freedom to do whatever they want to their forked version (aka. derivative work), which includes publishing on AppStore, etc. If a copyleft license is choiced (I assume GPLv2 here just for example), *these freedoms (in my definition)* will be taken from those developers.

Actually, I already see a lot of projects that uses a copylefted license but get abandoned and nobody continues to maintain it. So you see copyleft license can not be a guarantee of the project lifecycle and development.

And I also would like to make it clear, the (free software) license is always applied to the source code (or binary) you've already gotten, it's not applied to the whole project lifecycle.

----------

As a free software advocator, I do think it's necessary to tell users what is freedom in the software field. Thus these copywritings are there to do this job. Actually, in order to grants more freedoms to people, I also considering change it [1] to Expat/MIT and Apache 2.0 dual license model so more people can get benefits from the project.

[1]: At this moment, I'm still the solo copyright holder of this project's main (c++ and assets exclude translations other than en and zh_CN) codebase, so I can change the license at any time.

Since the four essential freedoms apply to all (FSF approved) free software license, that copywriting is always valid, so I insist to keep that text untouched. For the other copywriting issues not related to the license stuff, it may get a rewritten in later updates, I'll also leave it unchanged until it get rewritten since changing it will requires all translation contributors redo their work.

I'd like to thank you for the contributon again here whatever this patch get merged or not, if you have any advice about the correct way to introduce software freedom to user, feel free to start a new discussion in the discussion page :)

**Pull request closed without merging**
